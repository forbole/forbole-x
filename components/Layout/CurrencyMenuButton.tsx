import { Button, Menu, MenuItem } from '@material-ui/core';
import React from 'react';
import currencies from '../../misc/currencies';
import useStyles from './styles';
import CurrencyIcon from '../../assets/images/icons/icon_delegate_08.svg';
import { useGeneralContext } from '../../contexts/GeneralContext';
import useIconProps from '../../misc/useIconProps';

const CurrencyMenuButton: React.FC = () => {
  const { currency, setCurrency } = useGeneralContext();
  const classes = useStyles({});
  const iconProps = useIconProps(3);

  const [currencyAnchor, setCurrencyAnchor] = React.useState<Element>();

  const onClose = React.useCallback(() => setCurrencyAnchor(undefined), [setCurrencyAnchor]);
  const onItemClick = React.useCallback(
    c => {
      setCurrency(c);
      onClose();
    },
    [setCurrency, onClose],
  );

  return (
    <>
      <Button
        onClick={e => setCurrencyAnchor(e.currentTarget)}
        className={classes.navBarButton}
        startIcon={<CurrencyIcon {...iconProps} />}
      >
        {currency}
      </Button>
      <Menu
        anchorEl={currencyAnchor}
        getContentAnchorEl={null}
        anchorOrigin={{
          vertical: 'bottom',
          horizontal: 'center',
        }}
        transformOrigin={{
          vertical: 'top',
          horizontal: 'center',
        }}
        keepMounted
        open={!!currencyAnchor}
        onClose={onClose}
      >
        {currencies
          .filter(c => c !== currency)
          .map(c => (
            <MenuItem key={c} onClick={() => onItemClick(c)}>
              {c}
            </MenuItem>
          ))}
      </Menu>
    </>
  );
};

export default React.memo(CurrencyMenuButton);
