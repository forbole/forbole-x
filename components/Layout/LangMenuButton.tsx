import { Button, Menu, MenuItem } from '@material-ui/core';
import React from 'react';
import Link from 'next/link';
import useTranslation from 'next-translate/useTranslation';
import { useRouter } from 'next/router';
import useStyles from './styles';
import LangIcon from '../../assets/images/icons/icon_language.svg';
import useIconProps from '../../misc/useIconProps';

const LangMenuButton: React.FC = () => {
  const { t, lang } = useTranslation('common');
  const { locales, pathname, query } = useRouter();
  const iconProps = useIconProps(3);
  const classes = useStyles({});
  const [anchor, setAnchor] = React.useState<Element>();

  const onClose = React.useCallback(() => setAnchor(undefined), [setAnchor]);

  React.useEffect(() => {
    onClose();
  }, [lang]);

  return (
    <>
      <Button
        onClick={e => setAnchor(e.currentTarget)}
        className={classes.navBarButton}
        startIcon={<LangIcon {...iconProps} />}>
        {t(lang)}
      </Button>
      <Menu
        anchorEl={anchor}
        getContentAnchorEl={null}
        anchorOrigin={{
          vertical: 'bottom',
          horizontal: 'center',
        }}
        transformOrigin={{
          vertical: 'top',
          horizontal: 'center',
        }}
        keepMounted
        open={!!anchor}
        onClose={onClose}>
        {locales
          .filter(l => l !== lang)
          .map(l => (
            <div key={l}>
              <Link
                href={{
                  pathname,
                  query,
                }}
                locale={l}
                passHref>
                <MenuItem button component="a">
                  {t(l)}
                </MenuItem>
              </Link>
            </div>
          ))}
      </Menu>
    </>
  );
};

export default React.memo(LangMenuButton);
