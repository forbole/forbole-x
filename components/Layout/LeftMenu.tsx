import React from 'react'
import useTranslation from 'next-translate/useTranslation'
import { Box, List, ListItem, ListItemIcon, ListItemText, Paper, useTheme } from '@material-ui/core'
import Link from 'next/link'
import OverviewIcon from '../../assets/images/icons/icon_overview.svg'
import WalletManageIcon from '../../assets/images/icons/icon_wallet_manage.svg'
import DelegateIcon from '../../assets/images/icons/icon_delegate_08.svg'
import MarketIcon from '../../assets/images/icons/icon_market.svg'
import ProposalIcon from '../../assets/images/icons/icon_proposal.svg'
import AddressBookIcon from '../../assets/images/icons/icon_address_book.svg'
import ExplorerIcon from '../../assets/images/icons/icon_explorer.svg'
import Logo from '../../assets/images/logo.svg'
import LogoExpended from '../../assets/images/logo_expended.svg'
import useStyles from './styles'
import useIconProps from '../../misc/useIconProps'
import { MenuWidth } from '.'
import { useGeneralContext } from '../../contexts/GeneralContext'
import { CustomTheme } from '../../misc/theme'

interface LeftMenuProps {
  activeItem: string
  isMenuExpanded: boolean
  setIsMenuExpanded(isExpanded: boolean): void
}

const LeftMenu: React.FC<LeftMenuProps> = ({ activeItem, isMenuExpanded, setIsMenuExpanded }) => {
  const { t } = useTranslation('common')
  const themeStyle: CustomTheme = useTheme()
  const iconProps = useIconProps(3)
  const classes = useStyles()
  const { theme } = useGeneralContext()
  const items = React.useMemo(
    () => [
      {
        title: t('overview'),
        icon: <OverviewIcon {...iconProps} />,
        href: '/',
      },
      {
        title: t('wallet manage'),
        icon: <WalletManageIcon {...iconProps} />,
        href: '/wallets',
      },
      {
        title: t('delegate'),
        icon: <DelegateIcon {...iconProps} />,
        href: '/delegate',
      },
      // {
      //   title: t('market'),
      //   icon: <MarketIcon {...iconProps} />,
      //   href: '/market',
      // },
      {
        title: t('proposals'),
        icon: <ProposalIcon {...iconProps} />,
        href: '/proposals',
      },
      // {
      //   title: t('address book'),
      //   icon: <AddressBookIcon {...iconProps} />,
      //   href: '/address-book',
      // },
      // {
      //   title: t('explorer'),
      //   icon: <ExplorerIcon {...iconProps} />,
      //   href: '/explorer',
      // },
    ],
    [iconProps, t]
  )

  return (
    <Paper
      style={{
        width: themeStyle.spacing(isMenuExpanded ? MenuWidth.Expanded : MenuWidth.Collapsed),
      }}
      className={classes.leftMenuContainer}
    >
      <List className={classes.menu}>
        <ListItem
          className={classes.menuItem}
          button
          onClick={() => setIsMenuExpanded(!isMenuExpanded)}
        >
          <ListItemIcon>
            <Box ml={-1}>
              {isMenuExpanded ? (
                <LogoExpended
                  fill={
                    theme === 'light'
                      ? themeStyle.palette.primary.main
                      : themeStyle.palette.text.primary
                  }
                />
              ) : (
                <Logo
                  fill={
                    theme === 'light'
                      ? themeStyle.palette.primary.main
                      : themeStyle.palette.text.primary
                  }
                />
              )}
            </Box>
          </ListItemIcon>
        </ListItem>
        {items.map((item) => {
          const selected = item.href === activeItem
          return (
            <Link key={item.title} href={item.href} passHref>
              <ListItem
                selected={selected}
                className={classes.menuItem}
                button
                component="a"
                style={{ background: selected ? themeStyle.palette.menuBackground : 'inherits' }}
              >
                <ListItemIcon>
                  {React.cloneElement(item.icon, {
                    fill: selected ? themeStyle.palette.primary.main : themeStyle.palette.grey[300],
                  })}
                </ListItemIcon>
                <ListItemText
                  primary={item.title}
                  primaryTypographyProps={{
                    variant: 'h6',
                    color: selected ? 'primary' : 'textSecondary',
                  }}
                />
              </ListItem>
            </Link>
          )
        })}
      </List>
    </Paper>
  )
}

export default React.memo(LeftMenu)
