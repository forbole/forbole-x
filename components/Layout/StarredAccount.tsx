import React from 'react';
import Link from 'next/link';
import useTranslation from 'next-translate/useTranslation';
import {
  Avatar,
  Box,
  ListItem,
  ListItemIcon,
  ListItemText,
  useTheme,
  CircularProgress,
  Typography,
} from '@material-ui/core';
import { useGeneralContext } from '../../contexts/GeneralContext';
import {
  formatTokenAmount,
  getTotalBalance,
  getTotalTokenAmount,
  transformGqlAcountBalance,
} from '../../misc/utils';
import { CustomTheme } from '../../misc/theme';
import cryptocurrencies from '../../misc/cryptocurrencies';
import useStyles from './styles';
import useLatestAccountBalance from '../../graphql/hooks/useLatestAccountBalance';

interface StarredAccountProps {
  account: Account;
}

const StarredAccount: React.FC<StarredAccountProps> = ({ account }) => {
  const themeStyle: CustomTheme = useTheme();
  const { lang } = useTranslation('common');
  const { hideAmount } = useGeneralContext();
  const classes = useStyles({});
  const crypto = cryptocurrencies[account.crypto];
  // Latest data
  const { data: balanceData, loading } = useLatestAccountBalance(account.crypto, account.address);
  const { tokenAmounts } = React.useMemo(() => {
    const accountBalance = transformGqlAcountBalance(balanceData, Date.now());
    return {
      tokenAmounts: getTotalTokenAmount(accountBalance).amount,
      usdBalance: getTotalBalance(accountBalance).balance,
    };
  }, [balanceData]);
  return (
    <Link key={account.address} href="/account/[address]" as={`/account/${account.address}`}>
      <ListItem className={classes.favMenuItem} button component="a">
        <ListItemIcon>
          <Avatar
            alt={crypto.name}
            src={crypto.image}
            style={{ height: themeStyle.spacing(4), width: themeStyle.spacing(4) }}
          />
        </ListItemIcon>
        <Box display="flex" flexDirection="column">
          <ListItemText
            primary={account.name}
            primaryTypographyProps={{
              variant: 'h6',
            }}
          />
          {loading ? (
            <Box>
              <CircularProgress size={20} />
            </Box>
          ) : (
            <Box mt={-1}>
              <Typography variant="body2" color="textSecondary">
                {formatTokenAmount(tokenAmounts, { defaultUnit: account.crypto, lang, hideAmount })}
              </Typography>
              {/* after USD balance is added */}
              {/* <Typography>{formatCurrency(usdBalance, currency, lang)}</Typography> */}
            </Box>
          )}
        </Box>
      </ListItem>
    </Link>
  );
};

export default StarredAccount;
