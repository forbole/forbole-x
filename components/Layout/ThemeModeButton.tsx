import { IconButton } from '@material-ui/core';
import React from 'react';
import useStyles from './styles';
import DarkModeIcon from '../../assets/images/icons/icon_dark_mode.svg';
import LightModeIcon from '../../assets/images/icons/icon_light_mode.svg';
import { useGeneralContext } from '../../contexts/GeneralContext';
import useIconProps from '../../misc/useIconProps';

const ThemeModeButton: React.FC = () => {
  const { theme, setTheme } = useGeneralContext();
  const iconProps = useIconProps(3);
  const classes = useStyles({});
  const isDarkMode = theme === 'dark';
  return (
    <IconButton
      onClick={() => setTheme(isDarkMode ? 'light' : 'dark')}
      className={classes.navBarButton}
    >
      {isDarkMode ? <LightModeIcon {...iconProps} /> : <DarkModeIcon {...iconProps} />}
    </IconButton>
  );
};

export default React.memo(ThemeModeButton);
