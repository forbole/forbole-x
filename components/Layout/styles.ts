import { makeStyles } from '@material-ui/core/styles';
import { CustomTheme } from '../../misc/theme';

interface StyleProps {
  mode?: string;
}

const useStyles = makeStyles(
  (theme: CustomTheme) => ({
    leftMenuContainer: {
      position: 'fixed',
      top: 0,
      bottom: 0,
      left: 0,
      overflow: 'hidden',
      overflowX: 'hidden',
      zIndex: 2,
      transition: 'width 0.2s ease-in-out',
    },
    navbarLeftComponentContainer: {
      transition: 'margin-left 0.2s ease-in-out',
    },
    menu: {
      margin: theme.spacing(2),
      display: 'flex',
      flexDirection: 'column',
      height: '100%',
    },
    menuItems: {
      height: '60%',
      overflowY: 'auto',
      overflowX: 'hidden',
    },
    menuItem: {
      height: theme.spacing(6),
      paddingLeft: theme.spacing(1.5),
      paddingRight: theme.spacing(1.5),
      marginBottom: theme.spacing(3),
      borderRadius: theme.spacing(1),
      whiteSpace: 'nowrap',
    },
    favMenu: {
      height: '40%',
      overflowY: 'auto',
      overflowX: 'hidden',
      marginBottom: theme.spacing(4),
    },
    starredAccounts: {
      transition: 'display 0.2s ease-in-out',
      height: theme.spacing(6),
      marginBottom: theme.spacing(3),
      borderRadius: theme.spacing(1),
      padding: theme.spacing(2),
      paddingBottom: 0,
      whiteSpace: 'nowrap',
    },
    manageAccounts: {
      transition: 'display 0.2s ease-in-out',
      height: theme.spacing(6),
      marginBottom: theme.spacing(3),
      borderRadius: theme.spacing(1),
      padding: theme.spacing(2),
      paddingTop: 0,
      overflowWrap: 'break-word',
      wordWrap: 'break-word',
      hyphens: 'auto',
    },
    starButton: {
      marginTop: theme.spacing(2),
      marginLeft: theme.spacing(2),
      paddingLeft: theme.spacing(3),
      paddingRight: theme.spacing(3),
      width: 'fit-content',
    },
    favMenuItem: {
      height: theme.spacing(8),
      borderRadius: theme.spacing(1),
      whiteSpace: 'nowrap',
      paddingLeft: theme.spacing(1.5),
      paddingRight: theme.spacing(1.5),
    },
    navBar: {
      position: 'fixed',
      top: 0,
      right: 0,
      left: 0,
      height: theme.spacing(11),
      display: 'flex',
      justifyContent: 'flex-end',
      alignItems: 'center',
      padding: theme.spacing(3),
      backgroundColor: theme.palette.background.default,
      zIndex: 1,
    },
    navBarButton: {
      marginLeft: theme.spacing(2),
    },
    main: {
      padding: theme.spacing(3),
      paddingTop: 0,
      marginTop: theme.spacing(11),
      display: 'flex',
      flexDirection: 'column',
      transition: 'margin-left 0.2s ease-in-out',
    },
    mobileSignoutButton: {
      margin: theme.spacing(3),
      position: 'absolute',
      bottom: 0,
      left: 0,
      right: 0,
    },
    arrow: {
      '&:before': {
        border: (props: StyleProps) =>
          props.mode === 'light' ? `1px solid ${theme.palette.text.secondary}` : 'none',
      },
      color: theme.palette.dialogBackground,
    },
    tooltip: {
      backgroundColor: theme.palette.dialogBackground,
      border: (props: StyleProps) =>
        props.mode === 'light' ? `1px solid ${theme.palette.text.secondary}` : 'none',
      color: theme.palette.text.secondary,
      fontSize: theme.spacing(1.8),
      padding: theme.spacing(1, 1.5),
    },
  }),
  {
    name: 'HookGlobalStyles',
    index: 2,
  },
);

export default useStyles;
