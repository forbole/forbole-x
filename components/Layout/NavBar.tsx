import {
  Box,
  Button,
  Drawer,
  IconButton,
  List,
  ListItem,
  ListItemIcon,
  ListItemText,
  useTheme,
} from '@material-ui/core';
import React from 'react';
import Link from 'next/link';
import useTranslation from 'next-translate/useTranslation';
import { useRouter } from 'next/router';
import useStyles from './styles';
// import NotiIcon from '../../assets/images/icons/icon_notification_71.svg'
import SettingsIcon from '../../assets/images/icons/icon_settings.svg';
import AddressBookIcon from '../../assets/images/icons/icon_address_book.svg';
import MenuIcon from '../../assets/images/icons/icon_menu.svg';
import BackIcon from '../../assets/images/icons/icon_back.svg';
import WalletManageIcon from '../../assets/images/icons/icon_wallet_manage.svg';
import HideIcon from '../../assets/images/icons/icon_hide.svg';
import ShowIcon from '../../assets/images/icons/icon_show.svg';
import LogOutIcon from '../../assets/images/icons/log-out.svg';
import CurrencyMenuButton from './CurrencyMenuButton';
// import LangMenuButton from './LangMenuButton'
import ThemeModeButton from './ThemeModeButton';
import useIconProps from '../../misc/useIconProps';
import { useWalletsContext } from '../../contexts/WalletsContext';
import { useGeneralContext } from '../../contexts/GeneralContext';

const NavBar: React.FC<{
  HeaderLeftComponent?: React.ReactNode;
  ChromeExtTitleComponent?: React.ReactNode;
  ChromeExtRightComponent?: React.ReactNode;
  back?: boolean;
  menuWidth: number;
  isChromeExt: boolean;
}> = ({
  HeaderLeftComponent,
  ChromeExtTitleComponent,
  ChromeExtRightComponent,
  menuWidth,
  isChromeExt,
  back,
}) => {
  const iconProps = useIconProps(3);
  const classes = useStyles({});
  const { t } = useTranslation('common');
  const theme = useTheme();
  const router = useRouter();
  const { signOut } = useWalletsContext();
  const { setHideAmount, hideAmount } = useGeneralContext();
  const [isChromeExtMenuOpen, setIsChromeExtMenuOpen] = React.useState(false);

  const chromeExtMenuItems = React.useMemo(
    () => [
      {
        title: t('wallet'),
        icon: <WalletManageIcon {...iconProps} />,
        href: '/wallets',
      },
      {
        title: t('settings'),
        icon: <SettingsIcon {...iconProps} />,
        href: '/settings',
      },
      {
        title: t('address book'),
        icon: <AddressBookIcon {...iconProps} />,
        href: '/address-book',
        // divider: true,
      },
      // {
      //   title: t('support'),
      //   href: '/settings?tab=support',
      // },
      // {
      //   title: t('feedback'),
      //   href: '/settings?tab=feedback',
      // },
      // {
      //   title: t('about'),
      //   href: '/settings?tab=about',
      // },
    ],
    [iconProps, t],
  );

  return isChromeExt ? (
    <>
      <Box
        className={classes.navBar}
        alignItems="center"
        style={{ justifyContent: 'space-between' }}>
        <IconButton onClick={() => (back ? router.back() : setIsChromeExtMenuOpen(true))}>
          {back ? <BackIcon {...iconProps} /> : <MenuIcon {...iconProps} />}
        </IconButton>
        {ChromeExtTitleComponent || null}
        {ChromeExtRightComponent || <Box p={3} />}
      </Box>
      <Drawer
        anchor="left"
        open={isChromeExtMenuOpen}
        onClose={() => setIsChromeExtMenuOpen(false)}>
        <Box display="flex" justifyContent="flex-end" m={2}>
          <ThemeModeButton />
        </Box>
        <List style={{ width: theme.spacing(menuWidth), padding: theme.spacing(0, 2) }} id="navBar">
          {chromeExtMenuItems.map(item => {
            return (
              <React.Fragment key={item.title}>
                <Link href={item.href} passHref>
                  <ListItem className={classes.menuItem} button component="a">
                    {item.icon ? <ListItemIcon>{item.icon}</ListItemIcon> : null}
                    <ListItemText
                      primary={item.title}
                      primaryTypographyProps={{
                        variant: 'h6',
                        color: 'textSecondary',
                      }}
                    />
                  </ListItem>
                </Link>
                {/* {item.divider ? (
                  <Box mb={4}>
                    <Divider />
                  </Box>
                ) : null} */}
              </React.Fragment>
            );
          })}
        </List>
        <Button
          className={classes.mobileSignoutButton}
          variant="outlined"
          startIcon={<LogOutIcon {...iconProps} />}
          onClick={signOut}>
          {t('sign out')}
        </Button>
      </Drawer>
    </>
  ) : (
    <Box className={classes.navBar} id="navBar">
      {HeaderLeftComponent ? (
        <Box className={classes.navbarLeftComponentContainer} flex={1} ml={menuWidth}>
          {HeaderLeftComponent}
        </Box>
      ) : null}
      <CurrencyMenuButton />
      {/* <LangMenuButton />
      <IconButton className={classes.navBarButton}>
        <NotiIcon {...iconProps} />
      </IconButton> */}
      <IconButton className={classes.navBarButton} onClick={() => setHideAmount(h => !h)}>
        {hideAmount ? <HideIcon {...iconProps} /> : <ShowIcon {...iconProps} />}
      </IconButton>
      <ThemeModeButton />
      <Link href="/settings" passHref>
        <IconButton className={classes.navBarButton}>
          <SettingsIcon {...iconProps} />
        </IconButton>
      </Link>
      <Button
        className={classes.navBarButton}
        startIcon={<LogOutIcon {...iconProps} />}
        onClick={signOut}>
        {t('sign out')}
      </Button>
    </Box>
  );
};

export default React.memo(NavBar);
