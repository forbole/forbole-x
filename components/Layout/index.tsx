import React from 'react';
import { Box, useTheme } from '@material-ui/core';
import { useRouter } from 'next/router';
import get from 'lodash/get';
import qs from 'query-string';
import lzutf8 from 'lzutf8';
import useStyles from './styles';
import LeftMenu from './LeftMenu';
import NavBar from './NavBar';
import GetStarted from '../GetStarted';
import { useWalletsContext } from '../../contexts/WalletsContext';
import UnlockPasswordDialog from '../UnlockPasswordDialog';
import usePersistedState from '../../misc/usePersistedState';
import ConfirmTransactionDialog from '../ConfirmTransactionDialog';
import ChromeExtDialog from '../ChromeExtDialog';
import useIsChromeExt from '../../misc/useIsChromeExt';

export enum MenuWidth {
  Expanded = 32,
  Collapsed = 10,
}

interface LayoutProps {
  activeItem?: string;
  passwordRequired?: boolean;
  children: React.ReactNode;
  HeaderLeftComponent?: React.ReactNode;
  ChromeExtTitleComponent?: React.ReactNode;
  ChromeExtRightComponent?: React.ReactNode;
  back?: boolean;
}

const Layout: React.FC<LayoutProps> = ({
  activeItem,
  passwordRequired,
  HeaderLeftComponent,
  ChromeExtTitleComponent,
  ChromeExtRightComponent,
  back,
  children,
}) => {
  const classes = useStyles({});
  const theme = useTheme();
  const [isMenuExpanded, setIsMenuExpanded, loaded] = usePersistedState('isMenuExpanded', true);
  const { isFirstTimeUser, appUnlockState, isChromeExtInstalled, unlockWallets, password } =
    useWalletsContext();
  // Hide menu for chrome extension
  const router = useRouter();
  const { isChromeExt } = useIsChromeExt();

  // Open ConfirmTransactionDialog with correct query params
  const { address, transactionData, open, onClose } = React.useMemo(() => {
    const { url, query } = qs.parseUrl(router.asPath);
    return {
      address: get(router, 'query.address', ''),
      transactionData: get(router, 'query.transactionData', '')
        ? JSON.parse(
            lzutf8.decompress(get(router, 'query.transactionData', '""'), {
              inputEncoding: 'Base64',
            }),
          )
        : '',
      open: !!get(router, 'query.transactionData', ''),
      onClose: () =>
        router.replace(
          qs.stringifyUrl({
            url,
            query: { ...query, transactionData: undefined, address: undefined },
          }),
        ),
    };
  }, [router]);

  const initPassword = React.useCallback(async () => {
    try {
      if (appUnlockState !== 'unlocked') {
        if (localStorage.getItem('web-unlock-pasword')) {
          await unlockWallets(String(localStorage.getItem('web-unlock-pasword')));
          localStorage.removeItem('web-unlock-pasword');
        } else if (password) {
          await unlockWallets(password);
        }
      }
    } catch (err) {
      console.log(err);
    }
  }, [password, appUnlockState, unlockWallets]);

  React.useEffect(() => {
    initPassword();
  }, [initPassword]);

  return loaded ? (
    <>
      <NavBar
        HeaderLeftComponent={HeaderLeftComponent}
        ChromeExtTitleComponent={ChromeExtTitleComponent}
        ChromeExtRightComponent={ChromeExtRightComponent}
        // eslint-disable-next-line no-nested-ternary
        menuWidth={isChromeExt || isMenuExpanded ? MenuWidth.Expanded : MenuWidth.Collapsed}
        isChromeExt={isChromeExt}
        back={back}
      />
      {isChromeExt ? null : (
        <LeftMenu
          activeItem={activeItem}
          isMenuExpanded={isMenuExpanded}
          setIsMenuExpanded={setIsMenuExpanded}
        />
      )}
      <Box
        className={classes.main}
        style={{
          marginLeft: isChromeExt
            ? 0
            : theme.spacing(isMenuExpanded ? MenuWidth.Expanded : MenuWidth.Collapsed),
        }}>
        {passwordRequired && isFirstTimeUser ? <GetStarted /> : null}
        {!passwordRequired || appUnlockState === 'unlocked' ? children : null}
      </Box>
      {passwordRequired && !isFirstTimeUser && isChromeExtInstalled ? (
        <UnlockPasswordDialog />
      ) : null}
      {!isChromeExtInstalled ? <ChromeExtDialog /> : null}
      {open ? (
        <ConfirmTransactionDialog
          address={address as string}
          transactionData={transactionData}
          open={open}
          onClose={onClose}
        />
      ) : null}
    </>
  ) : null;
};

export default React.memo(Layout);
