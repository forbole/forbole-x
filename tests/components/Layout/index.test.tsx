import React from 'react'
import renderer from 'react-test-renderer'
import Layout from '../../../components/Layout'

const mockWalletsContext = {
  isFirstTimeUser: true,
  isChromeExtInstalled: true,
}

jest.mock('next/router', () => ({
  useRouter: () => ({
    asPath: 'https://localhost:3000',
  }),
}))
jest.mock('../../../contexts/WalletsContext', () => ({
  useWalletsContext: () => mockWalletsContext,
}))
jest.mock('../../../components/Layout/NavBar', () => 'NavBar')
jest.mock('../../../components/Layout/LeftMenu', () => (props) => <div id="LeftMenu" {...props} />)
jest.mock('../../../components/GetStarted', () => 'GetStarted')
jest.mock('../../../components/UnlockPasswordDialog', () => 'UnlockPasswordDialog')

describe('component: Layout', () => {
  it('renders default state for first time user correctly', () => {
    mockWalletsContext.isFirstTimeUser = true
    const component = renderer.create(
      <Layout activeItem="">
        <div />
      </Layout>
    )
    renderer.act(() => undefined)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
  it('renders passwordRequired state for first time user correctly', () => {
    mockWalletsContext.isFirstTimeUser = true
    const component = renderer.create(
      <Layout activeItem="" passwordRequired>
        <div />
      </Layout>
    )
    renderer.act(() => undefined)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
  it('renders default state for non first time user correctly', () => {
    mockWalletsContext.isFirstTimeUser = false
    const component = renderer.create(
      <Layout activeItem="">
        <div />
      </Layout>
    )
    renderer.act(() => undefined)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
  it('renders passwordRequired state for non first time user correctly', () => {
    mockWalletsContext.isFirstTimeUser = false
    const component = renderer.create(
      <Layout activeItem="" passwordRequired>
        <div />
      </Layout>
    )
    renderer.act(() => undefined)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
  it('renders menu collapsed state', () => {
    const component = renderer.create(
      <Layout activeItem="" passwordRequired>
        <div />
      </Layout>
    )
    renderer.act(() => undefined)
    renderer.act(() => {
      component.root.findByProps({ id: 'LeftMenu' }).props.setIsMenuExpanded(false)
    })
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})

afterEach(() => {
  jest.clearAllMocks()
})
