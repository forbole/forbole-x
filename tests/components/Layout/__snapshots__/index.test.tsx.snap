// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component: Layout renders default state for first time user correctly 1`] = `
Array [
  <NavBar
    menuWidth={32}
  />,
  <div
    activeItem=""
    id="LeftMenu"
    isMenuExpanded={true}
    setIsMenuExpanded={[Function]}
  />,
  <div
    className="MuiBox-root MuiBox-root-8 HookGlobalStyles-main-7"
    style={
      Object {
        "marginLeft": 256,
      }
    }
  >
    <div />
  </div>,
]
`;

exports[`component: Layout renders default state for non first time user correctly 1`] = `
Array [
  <NavBar
    menuWidth={32}
  />,
  <div
    activeItem=""
    id="LeftMenu"
    isMenuExpanded={true}
    setIsMenuExpanded={[Function]}
  />,
  <div
    className="MuiBox-root MuiBox-root-12 HookGlobalStyles-main-7"
    style={
      Object {
        "marginLeft": 256,
      }
    }
  >
    <div />
  </div>,
]
`;

exports[`component: Layout renders menu collapsed state 1`] = `
Array [
  <NavBar
    menuWidth={11.5}
  />,
  <div
    activeItem=""
    id="LeftMenu"
    isMenuExpanded={false}
    setIsMenuExpanded={[Function]}
  />,
  <div
    className="MuiBox-root MuiBox-root-14 HookGlobalStyles-main-7"
    style={
      Object {
        "marginLeft": 92,
      }
    }
  />,
  <UnlockPasswordDialog />,
]
`;

exports[`component: Layout renders passwordRequired state for first time user correctly 1`] = `
Array [
  <NavBar
    menuWidth={32}
  />,
  <div
    activeItem=""
    id="LeftMenu"
    isMenuExpanded={true}
    setIsMenuExpanded={[Function]}
  />,
  <div
    className="MuiBox-root MuiBox-root-11 HookGlobalStyles-main-7"
    style={
      Object {
        "marginLeft": 256,
      }
    }
  >
    <GetStarted />
  </div>,
]
`;

exports[`component: Layout renders passwordRequired state for non first time user correctly 1`] = `
Array [
  <NavBar
    menuWidth={32}
  />,
  <div
    activeItem=""
    id="LeftMenu"
    isMenuExpanded={true}
    setIsMenuExpanded={[Function]}
  />,
  <div
    className="MuiBox-root MuiBox-root-13 HookGlobalStyles-main-7"
    style={
      Object {
        "marginLeft": 256,
      }
    }
  />,
  <UnlockPasswordDialog />,
]
`;
